### PR TITLE
snap: add parsing support for hooks.

### PR DESCRIPTION
**Pull request by kyrofa**
This PR makes progress on LP: [#1586465](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1586465) by adding support for an optional "hooks" property in the YAML that follows the following format:

    hooks:
      hook-name:
        plugs: [plug-name]

Note that the plugs are optional. Also note that hooks are factored into the equation when considering whether or not a global plug is unbound-- if it's bound to either an app OR a hook it's considered bound for both, in which case the plug will not be bound to all apps and hooks. This is up for discussion.

Finally, note that `snap.InfoFromSnapYaml()` underwent a bit of a refactor here. It was just so long I was getting lost-- I simply broke it up into functional bits.

### CODE REVIEW

File: snap/validate.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ import (
 // Regular expression describing correct identifiers.
 var validName = regexp.MustCompile("^[a-z](?:-?[a-z0-9])*$")
 var validEpoch = regexp.MustCompile("^(?:0|[1-9][0-9]*[*]?)$")
+var validHookName = regexp.MustCompile(`^[a-z-]+$`)
```
Comment (niemeyer) [inline snap/validate.go]: This should be closer to `validName` above, as it prevents double dashes and dashes starting and ending the name.

File: snap/info_snap_yaml_test.go [hunk 2]
```diff
@@ -649,6 +687,212 @@ slots:
...
+		Hooks:     map[string]*snap.HookInfo{hook.Name: hook},
+	})
+	c.Check(hook, DeepEquals, &snap.HookInfo{
+		Snap:  info,
+		Name:  "test-hook",
+		Plugs: map[string]*snap.PlugInfo{plug.Name: plug},
+	})
+}
+
+func (s *YamlSuite) TestUnmarshalPlugsExplicitlyDefinedImplicitlyBoundToHooks(c *C) {
```
Comment (niemeyer) [inline snap/info_snap_yaml_test.go]: I suggest trying to shorten methods a bit, and using comments to explain the test with any space necessary.

**Comment by niemeyer** [comment]
Nice refactoring, thanks!

Only minor points. LGTM once you're happy about those.

File: snap/validate_test.go [hunk 3]
```diff
@@ -78,6 +78,33 @@ func (s *ValidateSuite) TestValidateEpoch(c *C) {
...
+		err := ValidateHook(hook)
+		c.Assert(err, IsNil)
+	}
+	invalidHooks := []*HookInfo{
+		&HookInfo{Name: ""},
+		&HookInfo{Name: "a a"},
+		&HookInfo{Name: "0"},
+		&HookInfo{Name: "123"},
+		&HookInfo{Name: "abc0"},
+		&HookInfo{Name: "日本語"},
```
Comment (mvo5) [inline snap/validate_test.go]: Adding to Gustavos point, maybe we should add here `a--a` and `-a` and `a-` once the validateHook gets udated.

**Comment by mvo5** [comment]
Looks good, nice work! I like the refactor as well. One (probably unneeded because obvious) comment about the hook name test. Otherwise I look forward for this to land.

File: snap/validate.go (hunk 1 above)
Comment (kyrofa) [inline snap/validate.go]: Done.

File: snap/validate_test.go (hunk 3 above)
Comment (kyrofa) [inline snap/validate_test.go]: Agreed, done.

File: snap/info_snap_yaml_test.go (hunk 2 above)
Comment (kyrofa) [inline snap/info_snap_yaml_test.go]: Alright they're shortened in the latest commit, though note that these were following the format of other tests in this file, so it still contains some ridiculously long ones.

**Pull request merged**